nano/node/fair_queue.hpp

```diff
@@ -0,0 +1,290 @@
+#pragma once
+
+#include <nano/lib/utility.hpp>
+#include <nano/node/transport/channel.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <deque>
+#include <functional>
+#include <memory>
+#include <numeric>
+#include <tuple>
+#include <utility>
+
+namespace nano
+{
+template <typename Request, typename Source>
+class fair_queue final
+{
+public:
+	/**
+	 * Holds user supplied source type(s) and an optional channel. This is used to uniquely identify and categorize the source of a request.
+	 */
+	struct origin
+	{
+		Source source;
+		std::shared_ptr<nano::transport::channel> channel;
+
+		origin (Source source, std::shared_ptr<nano::transport::channel> channel = nullptr) :
+			source{ source },
+			channel{ std::move (channel) }
+		{
+		}
+
+		bool alive () const
+		{
+			if (channel)
+			{
+				return channel->alive ();
+			}
+			// Some sources (eg. local RPC) don't have an associated channel, never remove their queue
+			return true;
+		}
+
+		auto operator<=> (origin const &) const = default;
+	};
+
+private:
+	struct entry
+	{
+		using queue_t = std::deque<Request>;
+		queue_t requests;
+
+		size_t priority;
+		size_t max_size;
+
+		entry (size_t max_size, size_t priority) :
+			priority{ priority },
+			max_size{ max_size }
+		{
+		}
+
+		Request pop ()
+		{
+			release_assert (!requests.empty ());
+
+			auto request = std::move (requests.front ());
+			requests.pop_front ();
+			return request;
+		}
+
+		bool push (Request request)
+		{
+			if (requests.size () < max_size)
+			{
+				requests.push_back (std::move (request));
+				return true; // Added
+			}
+			return false; // Dropped
+		}
+
+		bool empty () const
+		{
+			return requests.empty ();
+		}
+
+		size_t size () const
+		{
+			return requests.size ();
+		}
+	};
+
+public:
+	using source_type = origin;
```

Comment:
Can we rename source_type to origin_type, so we don't use two different names for the same thing (source/origin)?